src/Cake/Scripting/Roslyn/RoslynScriptSession.cs

```diff
@@ -78,62 +91,132 @@ public void ImportNamespace(string @namespace)
         {
-            // Generate the script code.
-            var generator = new RoslynCodeGenerator();
-            var code = generator.Generate(script);
-
-            // Warn about any code generation excluded namespaces
-            foreach (var @namespace in script.ExcludedNamespaces)
+            var scriptName = _options.Script.GetFilename();
+            using (var sw = new DisposableStopwatch(_log, string.Format("{0} execution time", scriptName)))
             {
-                _log.Warning("Namespace {0} excluded by code generation, affected methods:\r\n\t{1}",
-                    @namespace.Key, string.Join("\r\n\t", @namespace.Value));
-            }
-
-            // Create the script options dynamically.
-            var options = Microsoft.CodeAnalysis.Scripting.ScriptOptions.Default
-                .AddImports(Namespaces.Except(script.ExcludedNamespaces.Keys))
-                .AddReferences(References)
-                .AddReferences(ReferencePaths.Select(r => r.FullPath))
-                .WithEmitDebugInformation(_options.PerformDebug)
-                .WithMetadataResolver(Microsoft.CodeAnalysis.Scripting.ScriptMetadataResolver.Default);
-
-            var roslynScript = CSharpScript.Create(code, options, _host.GetType());
-
-            _log.Verbose("Compiling build script...");
-            var compilation = roslynScript.GetCompilation();
-            var diagnostics = compilation.GetDiagnostics();
-
-            var errors = new List<Diagnostic>();
-
-            foreach (var diagnostic in diagnostics)
-            {
-                switch (diagnostic.Severity)
+                var cacheDLLFileName = $"{scriptName}.dll";
+                var cacheHashFileName = $"{scriptName}.hash";
+                var cachedAssembly = _scriptCachePath.CombineWithFilePath(cacheDLLFileName).MakeAbsolute(_host.Context.Environment);
+                var hashFile = _scriptCachePath.CombineWithFilePath(cacheHashFileName);
+                string scriptHash = default;
+                if (_scriptCacheEnabled && IO.File.Exists(cachedAssembly.FullPath) && !_scriptForceRecompile)
                 {
-                    case DiagnosticSeverity.Info:
-                        _log.Information(diagnostic.ToString());
-                        break;
-                    case DiagnosticSeverity.Warning:
-                        _log.Warning(diagnostic.ToString());
-                        break;
-                    case DiagnosticSeverity.Error:
-                        _log.Error(diagnostic.ToString());
-                        errors.Add(diagnostic);
-                        break;
-                    default:
-                        break;
+                    _log.Verbose(cacheDLLFileName);
+                    scriptHash = FastHash.GenerateHash(Encoding.UTF8.GetBytes(string.Concat(script.Lines)));
+                    var cachedHash = IO.File.Exists(hashFile.FullPath) ? IO.File.ReadAllText(hashFile.FullPath) : string.Empty;
```

Comment:
Don't use `System.IO` use the `IFileSystem`

Reply:
Fixed